source/common/http/http1/codec_impl.cc

```diff
@@ -586,11 +586,19 @@ Http::Status ConnectionImpl::innerDispatch(Buffer::Instance& data) {
 
   ssize_t total_parsed = 0;
   if (data.length() > 0) {
-    for (const Buffer::RawSlice& slice : data.getRawSlices()) {
+    current_dispatching_buffer_ = &data;
+    while (data.length() > 0) {
+      auto slice = data.frontSlice();
+      dispatching_slice_already_drained_ = false;
       auto statusor_parsed = dispatchSlice(static_cast<const char*>(slice.mem_), slice.len_);
       if (!statusor_parsed.ok()) {
         return statusor_parsed.status();
       }
+      if (!dispatching_slice_already_drained_) {
+        ASSERT(statusor_parsed.value() <= slice.len_);
+        data.drain(statusor_parsed.value());
+      }
+
```

Comment:
This change looks amazing.  A possible comment is regarding tests; one possible test option would be to verify the slices that we expect to be moved are equal to the slices in the original buffer.